tools/fork-network/src/trim_database.rs

```diff
@@ -0,0 +1,159 @@
+        let epoch_id = final_head.epoch_id;
+        let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+        shard_layout.version()
+    };
+
+    // We'll parallel-iterate over the flat storage and also write to the new store at the same time.
+    // I'm not sure if there is value in parallelizing the write side as well, as that is now the
+    // bottleneck.
+    let (tx, rx) = std::sync::mpsc::sync_channel(1000000);
+    let iter_handle = {
+        let old = old.clone();
+        std::thread::spawn(move || {
+            StoreParallelIterator::for_each_in_range(
+                old,
+                DBCol::FlatState,
+                sharding_version.to_le_bytes().to_vec(),
+                (sharding_version + 1).to_le_bytes().to_vec(),
+                6,
+                move |key, value| {
+                    tx.send((key.to_vec(), value.to_vec())).unwrap();
+                },
+                true,
+            );
+        })
+    };
+
+    let mut update = new.store_update();
+    let mut non_inlined_keys = Vec::<Vec<u8>>::new();
+    let mut keys_copied = 0;
+    let mut bytes_copied = 0;
+    let mut bytes_this_batch = 0;
+    let mut last_print = std::time::Instant::now();
+    for (key, value) in rx {
+        update.set_raw_bytes(DBCol::FlatState, &key, &value);
+        keys_copied += 1;
+        bytes_copied += key.len() + value.len();
+        bytes_this_batch += key.len() + value.len();
+        if bytes_this_batch > 100 * 1024 * 1024 {
+            update.commit()?;
+            update = new.store_update();
+            bytes_this_batch = 0;
+        }
+
+        let value = FlatStateValue::try_from_slice(&value)?;
+        if let FlatStateValue::Ref(r) = value {
+            let (shard_uid, _) = decode_flat_state_db_key(&key).unwrap();
+            // This just happens to be the same key format as in the FlatState column.
+            non_inlined_keys.push(encode_flat_state_db_key(shard_uid, &r.hash.0));
+        }
+        if last_print.elapsed().as_secs() > 3 {
+            println!(
+                "FlatState: Copied {} keys, {} bytes, {} large keys to read from State",
+                keys_copied,
+                HumanBytes(bytes_copied as u64),
+                non_inlined_keys.len()
+            );
+            last_print = std::time::Instant::now();
+        }
+    }
+    update.commit()?;
+    println!(
+        "Done copying FlatState: Copied {} keys, {} bytes, {} large keys to read from State",
+        keys_copied,
+        HumanBytes(bytes_copied as u64),
+        non_inlined_keys.len()
+    );
+    iter_handle.join().unwrap();
+
+    // Now read and copy over the non-inlined keys.
+    let (tx, rx) = std::sync::mpsc::sync_channel(1000000);
+    let iter_handle = {
+        let old = old.clone();
+        std::thread::spawn(move || {
+            StoreParallelIterator::lookup_keys(
+                old,
+                DBCol::State,
+                non_inlined_keys,
```

Comment:
again I hit the same assert here because there were no non inlined keys. if you want to reproduce, what i did is i ran `python3 tests/sanity/transactions.py` with this change:

```diff
diff --git a/pytest/tests/sanity/transactions.py b/pytest/tests/sanity/transactions.py
index c628fbee0..a936e1e56 100755
--- a/pytest/tests/sanity/transactions.py
+++ b/pytest/tests/sanity/transactions.py
@@ -23,7 +23,7 @@ nodes = start_cluster(
     num_shards=4,
     config=None,
     genesis_config_changes=[["min_gas_price",
-                             0], ["max_inflation_rate", [0, 1]],
+                             0], ["use_production_config", True], ["max_inflation_rate", [0, 1]],
                             ["epoch_length", 10],
                             ["block_producer_kickout_threshold", 70]],
     client_config_changes={
@@ -33,7 +33,8 @@ nodes = start_cluster(
                     "secs": 2,
                     "nanos": 0
                 }
-            }
+            },
+            "tracked_shards": [0, 1, 2, 3]
         },
         1: {
             "consensus": {
@@ -41,7 +42,8 @@ nodes = start_cluster(
                     "secs": 2,
                     "nanos": 0
                 }
-            }
+            },
+            "tracked_shards": [0, 1, 2, 3]
         },
         2: {
             "consensus": {
@@ -49,7 +51,8 @@ nodes = start_cluster(
                     "secs": 2,
                     "nanos": 0
                 }
-            }
+            },
+            "tracked_shards": [0, 1, 2, 3]
         },
         3: {
             "consensus": {
@@ -57,7 +60,8 @@ nodes = start_cluster(
                     "secs": 2,
                     "nanos": 0
                 }
-            }
+            },
+            "tracked_shards": [0, 1, 2, 3]
         },
         4: {
             "consensus": {
@@ -72,7 +76,7 @@ nodes = start_cluster(
 
 started = time.time()
 
-act_to_val = [4, 4, 4, 4, 4]
+act_to_val = [0, 0, 0, 0, 0]
 
 ctx = utils.TxContext(act_to_val, nodes)
 
@@ -97,8 +101,12 @@ cur_balances = ctx.get_balances()
 assert cur_balances == ctx.expected_balances, "%s != %s" % (
     cur_balances, ctx.expected_balances)
 
+killed_it = False
 # we are done with the sanity test, now let's stress it
-for height, _ in utils.poll_blocks(nodes[4], timeout=TIMEOUT):
+for height, _ in utils.poll_blocks(nodes[0], timeout=TIMEOUT):
+    if height >= 60 and not killed_it:
+        nodes[4].kill()
+        killed_it=True
     if ctx.get_balances() == ctx.expected_balances:
         count = height - sent_height
         logger.info(f'Balances caught up, took {count} blocks, moving on')

```

and then after that test finished i ran:

```
$ python3 tests/mocknet/local_test_node.py local-test-setup --num-nodes 2 --neard-binary-path ~/nearcore/target/debug/neard --source-home-dir ~/.near/test0_finished --target-home-dir ~/.near/test0_finished --yes
$ python3 tests/mocknet/mirror.py --local-test new-test --epoch-length 200 --num-validators 2 --num-seats 100 --genesis-protocol-version 67 --yes
```

(if the new-test command fails with some timeout error, just try again, idk exactly why that happens). but then if you look at the logs in `~/.near/local-mocknet/node0/stderr`, youll see that the fork-network command panicked